packages/react-compose/src/next/getSlots.tsx

```diff
@@ -0,0 +1,52 @@
+import * as React from 'react';
+import { getNativeElementProps, omit } from '@uifabric/utilities';
+import { GenericDictionary } from './types';
+import { nullRender } from './nullRender';
+
+/**
+ * Given the state and an array of slot names, will break out `slots` and `slotProps`
+ * collections.
+ *
+ * The root is always derived from the `as` prop.
+ *
+ * Slots will render as null if they are rendered as primitives with undefined children.
+ *
+ * The slotProps will always omit the `as` prop within them, and for slots that are string
+ * primitives, the props will be filtered according the the slot type. For example, if the
+ * slot is rendered `as: 'a'`, the props will be filtered for acceptable anchor props.
+ *
+ * @param state - State including slot definitions
+ * @param slotNames - Name of which props are slots
+ * @returns An object containing the `slots` map and `slotProps` map.
+ */
+export const getSlots = (state: GenericDictionary, slotNames?: string[] | undefined) => {
+  const slots: GenericDictionary = {
+    root: state.as || nullRender,
+  };
+  const slotProps: GenericDictionary = {
+    root: typeof state.as === 'string' ? getNativeElementProps(state.as, state) : omit(state, { as: 1 }),
+  };
+
+  if (slotNames) {
+    for (const name of slotNames) {
+      const slotDefinition = state[name];
+      const { as: slotAs, children } = slotDefinition;
+      const isSlotPrimitive = typeof slotAs === 'string';
+
+      slots[name] = slotDefinition.children !== undefined || !isSlotPrimitive ? slotAs : nullRender;
```

Comment:
Maybe I do not understand the logic, but there can be component without children, but still valid for rendering, no?

```
slots: { divider: Divider },
slotProps: { divider: { vertical : true } },
```

How would this work?

Reply:
Only primitive slots are excluded; this is an optiization.

I found that most slots have some logic which says: put a slot here, contain it in a DIV, but if the user doesn't specify anything, don't render it. Something like this:

```jsx
const Foo = (props) => {
  const { slots, slotProps } = getSlots(props, [ 'icon' ]);
  return (
    <slots.root { ... }>
      { props.icon && <slots.icon /> }
    </slots.root>
  );
};
```

And actually in practice, hardly any dev avoids prepends the condition, so you end up with Buttons which have containers rendered containing nothing.

So I thought it would be wise to assume this default behavior: a slot rendered without content shouldn't be rendered.

Of course there are exceptions like `input` or `hr` or whatnot, which don't render children. For exceptions, you'd render null children:

```jsx
getSlots({ as: 'div', { divider: { as: 'hr', children: null }})
```

This will render the `hr` even if the user has not specified the `divider` slot.

Reply:
And just to reiterate; this exclusion behavior is only for primitives. For component containers, the slot is rendered no matter what, because the container would own how to manage the slot.

This is a perf optimization. If this just is too confusing, we can avoid. But trying to cull unneeded elements by default.

Reply:
I've opened an RFC here to track: https://github.com/microsoft/fluentui/issues/14374

Also scheduled some time next week. If this is too much magic, happy to revert.